File: articles/claude-shannon-and-the-mathematical-theory-of-information/prolog/entropy_example.pl
log2(X, Y) :- Y is log(X) / log(2).

entropy([], 0).
entropy([P|Ps], H) :-
    entropy(Ps, Rest),
    (P > 0 -> log2(P, L), H is Rest - P * L ; H is Rest).

% Query example:
% ?- entropy([0.5, 0.5], H).

File: articles/claude-shannon-and-the-mathematical-theory-of-information/cpp/entropy_example.cpp
#include <cmath>
#include <iostream>
#include <vector>

double entropy(const std::vector<double>& probs) {
    double total = 0.0;
    for (double p : probs) {
        if (p > 0.0) total += p * std::log2(p);
    }
    return -total;
}

int main() {
    std::cout << "entropy_bits=" << entropy({0.5, 0.5}) << "\n";
    return 0;
}
